src/target/rp.c

```diff
@@ -554,56 +611,216 @@ static void rp_spi_read(
 	}
 
 	/* Deselect the Flash and put things back to how they were */
-	rp_spi_chip_select(t, false);
+	rp_spi_chip_select(t, RP_GPIO_QSPI_CS_DRIVE_HIGH);
 	target_mem_write32(t, RP_SSI_ENABLE, 0);
 	target_mem_write32(t, RP_SSI_CTRL1, ctrl1);
 	target_mem_write32(t, RP_SSI_CTRL0, ctrl0);
 	target_mem_write32(t, RP_SSI_XIP_SPI_CTRL0, xpi_ctrl0);
 	target_mem_write32(t, RP_SSI_ENABLE, ssi_enabled);
 }
 
-static uint32_t rp_get_flash_length(target *t)
+// Connect the XIP controller to the flash pads
+static void rp_flash_connect_internal(target *t)
 {
-	uint32_t size = MAX_FLASH;
-	uint32_t bootsec[16];
-	size_t i;
-
-	rp_flash_resume(t);
-	target_mem_read(t, bootsec, RP_XIP_FLASH_BASE, sizeof(bootsec));
-	for (i = 0; i < 16; i++) {
-		if ((bootsec[i] != 0x00) && (bootsec[i] != 0xff))
-			break;
-	}
+    // Use hard reset to force IO and pad controls to known state (don't touch
+    // IO_BANK0 as that does not affect XIP signals)
+	uint32_t reset = target_mem_read32(t, RP_RESETS_RESET);
+    target_mem_write32(t, RP_RESETS_RESET, reset | RP_RESETS_RESET_IO_QSPI_BITS | RP_RESETS_RESET_PADS_QSPI_BITS);
+    target_mem_write32(t, RP_RESETS_RESET, reset);
+	while (~target_mem_read32(t, RP_RESETS_RESET_DONE) & (RP_RESETS_RESET_IO_QSPI_BITS | RP_RESETS_RESET_PADS_QSPI_BITS));
```

Comment:
~ is the wrong operation here - should be ! or an explicit comparison. Also, please use `continue` on a new line as the loop body so the loop body is obvious and explicit. A ';' on the end is just too easy to miss.

Reply:
It is the correct operation... We are bitwise inverting the value in register `RP_RESETS_RESET_DONE` and then bitwise and to the `IO_QSPI_BITS` and `PADS_QSPI_BITS`. I will re-write this to help improve clarity.

Reply:
Fixed!